src/main/scala/com/cognite/sdk/scala/common/GzipSttpBackend.scala

```diff
@@ -0,0 +1,109 @@
+package com.cognite.sdk.scala.common
+
+import java.io.ByteArrayOutputStream
+import java.nio.charset.Charset
+import java.util.zip.GZIPOutputStream
+
+import com.softwaremill.sttp._
+import org.apache.commons.io.IOUtils
+
+class GzipSttpBackend[R[_], S](delegate: SttpBackend[R, S], val minimumSize: Int = 1000)
+    extends SttpBackend[R, S] {
+  import GzipSttpBackend._
+
+  @SuppressWarnings(Array("org.wartremover.warts.Equals"))
+  implicit final class AnyOps[A](self: A) {
+    def strictEq(other: A): Boolean = self == other
+  }
+
+  private def isGzipped(keyAndValue: Tuple2[String, String]) =
+    keyAndValue._1.equalsIgnoreCase(HeaderNames.ContentEncoding) &&
+      keyAndValue._2.equalsIgnoreCase("gzip")
+
+  private def isGzippedContent(keyAndValue: Tuple2[String, String]) =
+    keyAndValue._1.equalsIgnoreCase(HeaderNames.ContentType) &&
+      keyAndValue._2.toLowerCase.contains("/gzip")
+
+  override def send[T](request: Request[T, S]): R[Response[T]] = {
+    val (newBody, newContentLength) = request.body match {
+      case body: BasicRequestBody
+          if !request.headers.exists(isGzipped)
```

Comment:
Out of interest: How could already zipped requests end up here?

Reply:
It might be overly paranoid, but as a generic middleware it shouldn't make assumptions about what other middlewares might have done to the request.

Messing up the order between the retry middleware and this one could for example result in multiple compressions if we didn't have this check (such an order would be a mistake, and should be fixed either way, but we don't have to make it horrendous).